doc/sphinx/proof-engine/ltac.rst

```diff
@@ -408,221 +412,242 @@ We can branch with the following structure:
+
+   :tacn:`exactly_once` is an :token:`l3_tactic`.
 
    .. warning::
 
       The experimental status of this tactic pertains to the fact if ``v``
-      performs side effects, they may occur in an unpredictable way. Indeed,
+      has side effects, they may occur in an unpredictable way. Indeed,
       normally ``v`` would only be executed up to the first success until
-      backtracking is needed, however exactly_once needs to look ahead to see
+      backtracking is needed, however :tacn:`exactly_once` needs to look ahead to see
       whether a second success exists, and may run further effects
       immediately.
 
    .. exn:: This tactic has more than one success.
       :undocumented:
 
-Checking the failure
-~~~~~~~~~~~~~~~~~~~~
+Checking for failure: assert_fails
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 Coq provides a derived tactic to check that a tactic *fails*:
 
-.. tacn:: assert_fails @ltac_expr
+.. tacn:: assert_fails @ltac_expr3
    :name: assert_fails
 
-   This behaves like :tacn:`idtac` if :n:`@ltac_expr` fails, and
-   behaves like :n:`fail 0 @ltac_expr "succeeds"` if :n:`@ltac_expr`
-   has at least one success.
+   If :n:`@ltac_expr3` fails, the proof state is unchanged and no message is printed.
+   If :n:`@ltac_expr3` unexpectedly has at least one success, the tactic performs
+   a :tacn:`gfail` :n:`0`, printing the following message:
 
-Checking the success
-~~~~~~~~~~~~~~~~~~~~
+   .. exn:: Tactic failure: <tactic closure> succeeds.
+      :undocumented:
+
+Checking for success: assert_succeeds
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 Coq provides a derived tactic to check that a tactic has *at least one*
 success:
 
-.. tacn:: assert_succeeds @ltac_expr
+.. tacn:: assert_succeeds @ltac_expr3
    :name: assert_succeeds
 
-   This behaves like
-   :n:`tryif (assert_fails @ltac_expr) then fail 0 @ltac_expr "fails" else idtac`.
+   If :n:`@ltac_expr3` has at least one success, the proof state is unchanged and
+   no message is printed.  If :n:`@ltac_expr3` fails, the tactic performs
+   a :tacn:`gfail` :n:`0`, printing the following message:
+
+   .. exn:: Tactic failure: <tactic closure> fails.
+      :undocumented:
 
 Solving
 ~~~~~~~
 
 We may consider the first to solve (i.e. which generates no subgoal)
 among a panel of tactics:
 
-.. tacn:: solve [{*| @ltac_expr}]
+.. tacn:: solve [ {*| @ltac_expr } ]
    :name: solve
 
    The :n:`@ltac_expr__i` are evaluated to :n:`v__i` and :n:`v__i` must be
-   tactic values, for i = 1, ..., n. Supposing n > 1,
-   :n:`solve [@ltac_expr__1 | ... | @ltac_expr__n]` applies :n:`v__1` to
+   tactic values, for i = 1, ..., n. If n > 1,
+   :tacn:`solve` :n:`[@ltac_expr__1 | ... | @ltac_expr__n]` applies :n:`v__1` to
    each goal independently and stops if it succeeds; otherwise it tries to
    apply :n:`v__2` and so on. It fails if there is no solving tactic.
 
+   :tacn:`solve` is an :token:`l1_tactic`.
+
    .. exn:: Cannot solve the goal.
```

Comment:
It looks like the actual error message is `No applicable tactic.`

Reply:
That message is already documented so it can't be repeated here.

Reply:
Right. But you could remove this incorrect one. And you could also refer the already documented message as is done here: https://github.com/coq/coq/blob/master/doc/sphinx/language/core/basic.rst#attributes (except that this example is referring to a warning). But that's up to you of course!